### PR TITLE
ARROW-10854: [Rust] [DataFusion] Simplify logical plan scans

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR simplifies the logical plan by removing `CsvScan`, `ParquetScan` and `InMemoryScan` and encapsulating all of them into `TableScan`.

The underlying aspect here is that all these nodes perform the same operation (through the `dyn TableProvider`), they are just described in the plan differently. This PR makes all scans be encapsulated via `TableScan`, thereby removing the other ones.

This has no execution differences - it only simplifies code at the logical level.

This also renames the function `MemTable::new -> Result<Self>` to `MemTable::try_new`, as that is the standard in rust for falible constructors.

### CODE REVIEW

**Comment by andygrove** [comment]
This is potentially going to be a breaking change for my use of the DataFusion logical plan in Ballista because I need to serialize logical plans to protobuf format to transmit them over the network. With the explicit variants for each type of input I have access to all the information I need, but with a reference to a `dyn TableProvider` this becomes opaque. I should have added tests for this use case.

I will have a think about how I can work around this as well. Maybe there are other approaches.

**Comment by jorgecarleitao** [comment]
@andygrove that is a good point that I had not thought about.

Unfortunately, I have also not though about them when reviewing the addition of `LogicalPlan::Extension`, `Expr::ScalarUDF`, and `Expr::AggregateUDF`, all of which directly or indirectly use `Arc<dyn X>`. Is this already a concern for Ballista?

Would it be possible to require `TableProvider` to implement [`serde::ser::Serialize`](https://docs.serde.rs/serde/ser/trait.Serialize.html), and use something like https://github.com/dtolnay/erased-serde to serialize trait objects?

Also relevant: https://stackoverflow.com/a/50026869/931303

**Comment by andygrove** [comment]
I think it would be helpful if I explained my design philosophy on this but also bear in mind that I am not an expert on this but this was my initial thinking:

- Logical plans should be trivially serializable. This is a good forcing function to make sure that they are easily constructed from any programming language and don't become Rust-specific. This is important when building bindings in other languages such as Python and Java and also for building drivers for distributed systems so that we can serialize query plans and send them to other processes.
- The execution context has registries of things like tables and udfs and is responsible for translating a logical plan into a physical plan, and it makes sense for the physical plan to use opaque dyn traits.

In Ballista, I have my own fork of the DataFusion logical plan currently, and I could continue with that but was considering using the DataFusion plan directly at some point (I'm not actively working on Ballista at the moment so this is all a bit theoretical).

I bet @alamb will have some valuable opinions on this.

File: rust/datafusion/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -71,37 +98,44 @@ impl LogicalPlanBuilder {
...
-        let p = ParquetTable::try_new(path)?;
-        let schema = p.schema();
+        let provider = Arc::new(ParquetTable::try_new(path)?);
+        let schema = provider.schema();
 
         let projected_schema = projection
-            .clone()
             .map(|p| Schema::new(p.iter().map(|i| schema.field(*i).clone()).collect()));
         let projected_schema = projected_schema.map_or(schema.clone(), SchemaRef::new);
+        let projected_schema = DFSchemaRef::new(DFSchema::from(&projected_schema)?);
```
Comment (alamb) [inline rust/datafusion/src/logical_plan/builder.rs:128]: As an aside I find the current conversion code to/from `Schema`/`DFSchema` quite cumbersome and hope to get a PR up later today to make it easier to use

**Review by alamb — APPROVED** [review]
I like where this is headed -- thank you @jorgecarleitao 

I think the ability to add a data source that is not InMemory, CSV or Parquet is 💯  and would be useful for anyone using DataFusion as a library. It is something I would likely need to suggest / propose shortly anyways as we will likely need something like this in IOx with some of our special storage engines. So thank you for taking the first step.

On serialization of LogicalPlans: I think being able to serialize / deserialize  to/from a stream of `byte`s is important for distributed execution (and potentially other things). I would personally suggest implementing something like `LogicalPlan::serialize` and `LogicalPlan::deseiralize` to keep the serialization specifics opaque to DataFusion users. With a little thought we could have some sort of extension mechanism to pass the right data along to user defined table sources to let systems like Ballista and IOx reconstitute the right `dyn` object. 

Investing in hooks to support user defined data sources seems like a ways to let people use DataFusion to build query systems on.

**Comment by andygrove** [comment]
I thought about this some more and I think that the serialization requirements I have in Ballista are beyond the scope of DataFusion since the goal is to support multiple query engines, with DataFusion being just one of them, so it makes sense for me to continue with Ballista having its own logical query plan that can be translated to DataFuision and other query engines. This change certainly improves the UX of using DataFusion so it looks good to me.

**Review by andygrove — APPROVED** [review]

File: rust/datafusion/src/logical_plan/builder.rs (hunk 1 above)
Comment (alamb) [reply rust/datafusion/src/logical_plan/builder.rs:128]: https://github.com/apache/arrow/pull/8888 for the record

**Pull request closed without merging**